[CI] MlMemoryIT testMemoryStats failing

**Build Scans:**
- [elasticsearch-intake #38134 / part4](https://gradle-enterprise.elastic.co/s/csr7j2d2b3bda)
- [elasticsearch-pull-request #139291 / part-4](https://gradle-enterprise.elastic.co/s/dr4c2pqxngym2)
- [elasticsearch-intake #38127 / part4](https://gradle-enterprise.elastic.co/s/b7ajcnmum6gwo)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:ml:qa:native-multi-node-tests:javaRestTest" --tests "org.elasticsearch.xpack.ml.integration.MlMemoryIT.testMemoryStats" -Dtests.seed=3C121BAEAE97F9FE -Dtests.locale=sr-BA -Dtests.timezone=Africa/Kinshasa -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.ml.integration.MlMemoryIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testMemoryStats'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: a collection with size <4>
     but: collection size was <3>
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testMemoryStats
- [main] 1 consecutive failures in step part4
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 3 failures in test testMemoryStats (0.7% fail rate in 448 executions)
- [main] 2 failures in step part4 (1.5% fail rate in 130 executions)
- [main] 2 failures in pipeline elasticsearch-intake (1.5% fail rate in 130 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -214,9 +214,6 @@ tests:
 - class: org.elasticsearch.index.mapper.IpOffsetDocValuesLoaderTests
   method: testOffsetArrayRandomHighCardinality
   issue: https://github.com/elastic/elasticsearch/issues/147086
-- class: org.elasticsearch.xpack.ml.integration.MlMemoryIT
-  method: testMemoryStats
-  issue: https://github.com/elastic/elasticsearch/issues/147122
 - class: org.elasticsearch.xpack.esql.tree.EsqlNodeSubclassTests
   method: testReplaceChildren {class org.elasticsearch.xpack.esql.plan.physical.ExchangeSinkExec}
   issue: https://github.com/elastic/elasticsearch/issues/147164

--- a/x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlMemoryIT.java
+++ b/x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/MlMemoryIT.java
@@ -7,6 +7,7 @@
 
 package org.elasticsearch.xpack.ml.integration;
 
+import org.elasticsearch.cluster.ClusterState;
 import org.elasticsearch.cluster.node.DiscoveryNodeRole;
 import org.elasticsearch.common.bytes.BytesArray;
 import org.elasticsearch.common.unit.ByteSizeValue;
@@ -36,6 +37,8 @@ import org.junit.After;
 
 import java.util.Base64;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 import static org.elasticsearch.xpack.ml.integration.ClassificationIT.KEYWORD_FIELD;
 import static org.elasticsearch.xpack.ml.integration.PyTorchModelIT.BASE_64_ENCODED_MODEL;
@@ -64,13 +67,22 @@ public class MlMemoryIT extends MlNativeDataFrameAnalyticsIntegTestCase {
         String dfaJobId = "dfa";
         startDataFrameAnalyticsJob(dfaJobId);
 
-        MlMemoryAction.Response response = client().execute(MlMemoryAction.INSTANCE, new MlMemoryAction.Request("_all")).actionGet();
+        int expectedNodes = cluster().size();
+        assertBusy(() -> {
+            ClusterState state = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState();
+            assertThat(state.nodes().getSize(), equalTo(expectedNodes));
+        }, 30, TimeUnit.SECONDS);
 
-        assertThat(response.failures(), empty());
-
+        AtomicReference<MlMemoryAction.Response> responseRef = new AtomicReference<>();
+        assertBusy(() -> {
+            MlMemoryAction.Response memoryResponse = client().execute(MlMemoryAction.INSTANCE, new MlMemoryAction.Request("_all"))
+                .actionGet();
+            assertThat(memoryResponse.failures(), empty());
+            assertThat(memoryResponse.getNodes(), hasSize(expectedNodes));
+            responseRef.set(memoryResponse);
+        }, 30, TimeUnit.SECONDS);
+        MlMemoryAction.Response response = responseRef.get();
         List<MlMemoryStats> statsList = response.getNodes();
-        // There are 4 nodes: 3 in the external cluster plus the test harness
-        assertThat(statsList, hasSize(4));
 
         int mlNodes = 0;
         int nodesWithPytorchModel = 0;

